### PR TITLE
Rewrite actions/modal and reducers/modal with typescript

### PR DESCRIPTION
**Pull request by takayamaki**
This PR rewrites actions/modal and reducers/modal with typescript and createAction.

### CODE REVIEW

File: app/javascript/mastodon/reducers/modal.ts
Comment (nschonni) [inline app/javascript/mastodon/reducers/modal.ts:77]: I don't think the extra ignore is required if you put an explanation string after the `@ts-expect-error`. The error should probably be addressed though

**Comment by renchap** [comment]
Thanks for continuing working on this <3

I will have a look, but I would like to merge #23631 before any other TS PRs are merged.

File: app/javascript/mastodon/reducers/modal.ts
Comment (renchap) [inline app/javascript/mastodon/reducers/modal.ts:59]: Can you rename it `modalReducer` so its more explicit?

**Review by renchap — CHANGES_REQUESTED** [review]
Nice work on this, just a small nitpick (and I think you need to rebase it, and re-run eslint with my recent changes?)

**Comment by renchap** [comment]
Also, could it be a good place to start rewriting the reducer with `createReducer` (or `createSlice`) from RTK?

This specific modal state seem to only be used in `modal_container.js`, so it can be easily converted into an Immer reducer.

If you feel confident with this, you can open another PR after this one doing it, or I can have a go as well. I would like a simple example of how a full rewrite to Typescript look like (Actions, Reducer, removing Immer, then rewriting into a FunctionComponent using Redux's hooks)

**Comment by takayamaki** [comment]
> Also, could it be a good place to start rewriting the reducer with `createReducer` (or `createSlice`) from RTK?
> 
> This specific modal state seem to only be used in `modal_container.js`, so it can be easily converted into an Immer reducer.
> 
> If you feel confident with this, you can open another PR after this one doing it, or I can have a go as well. I would like a simple example of how a full rewrite to Typescript look like (Actions, Reducer, removing Immer, then rewriting into a FunctionComponent using Redux's hooks)

OK, I will try that next week.
I attend RubyKaigi this week.

**Comment by takayamaki** [comment]
rebased.

I tried rewriting this reducer with createReducer.
However, it is need to rewrite [changeUploadCompose](https://github.com/mastodon/mastodon/blob/main/app/javascript/mastodon/actions/compose.js#L400) by createAsyncThunk without ts support.

It is difficult for me, and I seem that is too big to include this PR.

File: app/javascript/mastodon/reducers/index.ts
Comment (renchap) [inline app/javascript/mastodon/reducers/index.ts:53]: You should not change the key here, but use `modal: modalReducer`.

This would avoid changing the `modal` key everywhere this is accessed

**Review by renchap — CHANGES_REQUESTED** [review]
Looks good to me, except the `modalReducer` renaming in the state, it should be kept as `key`.

I missed that this was using async thunks, I will have a look later on the best way to handle this.

I think we could also have `OpenModalPayload` as a generic type, which would allow to properly type `modalProps`, I will try to find a way to make it work once this PR is merged.

Thanks again for this, good work!

File: app/javascript/mastodon/reducers/modal.ts
Comment (renchap) [reply app/javascript/mastodon/reducers/modal.ts:52]: This is not idiomatic, in this case you probably want to use `function()` with the correct argument & return types.

Both are pretty much the same (except the arrow function vs function).

Some prefer to extract the types, so it is easier to re-use them elsewhere without having to use `ReturnType<typeof myFunction>`, but I do not think we have a preferred syntax (yet) in the codebase.

Comment (ClearlyClaire) [reply app/javascript/mastodon/reducers/modal.ts:52]: I see. In this case I don't think this type would be reused elsewhere as it's pretty much an internal function. Therefore it'd make more sense to me to not name the type. That being said, I'm not that familiar with Typescript, so I'll let you decide on that.

Comment (takayamaki) [reply app/javascript/mastodon/reducers/modal.ts:52]: Often in Typescript, the first argument is an object.
This is similar to keyword argument in Ruby.
And JavaScript can [Destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) in function arguments.

In this case, there are some style for type annotation for function.

```typescript
type SomeFunction = (args: {
  str: string;
}) => string
const someFunction: SomeFunction = ({ str }) => {
  return(`some ${str}`)
}
```

```typescript
type SomeFunctionArg = {
  str: string;
};
const someFunction = ({ str }: SomeFunctionArg ): string => {
  return(`some ${str}`)
}
```

```typescript
const someFunction = ({ str }: {
  str: string;
}): string => {
  return(`some ${str}`)
}
```

IMO, I seems that 1st style are most simply for me.
Because typed variables accepts function object or not.

Comment (takayamaki) [reply app/javascript/mastodon/reducers/modal.ts:52]: I agree that it does not have to be the Arrow style.
However, when we use both of Arrow and `function` styles, a decision must be required about which to use **every time**.

If we use only an arrow style, there is no need to consider about it.
Tthis is an important advantage for newbie contributor friendly, I think.

In addition, I favourite that there is no confusion caused by `this` when use the arrow style.

I want to make these codes readable, mantainable, improvable for everyone even if not javascript specialist.
For it, We have to reduce confusion and cost of dicision as a possible.

Comment (takayamaki) [reply app/javascript/mastodon/reducers/modal.ts]: I rewrite type annotations into arrow function definitions.

@renchap Could you merge this PR?

**Review by renchap — APPROVED** [review]
Let's merge this, can you rebase one last time? 🙏

**Comment by takayamaki** [comment]
> Let's merge this, can you rebase one last time? 🙏

rebased!

**Pull request merged**
